### ISSUE TITLE
Feature: As an aiSSEMBLE user, I want to specify rules for my property values, so that configuration can be refreshed when needed

### ISSUE THREAD

**Issue by carter-cundiff** [boozallen/aissemble#118]
## Description
Follow on to implement #76. As part of Universal Configurations, we want to be able to specify rules for configuration properties within the vault server (like expiration date or use limits). This will allow the user to realize when property values are no longer valid and need to be updated.

## DOD
- [x] Create `propertyKey` class and refactor existing implemention to utilize it where applicable
- [x] Policies can be defined for properties outlining custom limitations (use limit, expiration date)
  - [x] Will be defined as JSON policies and include rules, regeneration method, and target properties
  - [x] Can define environment policies to override base policies
  - [x] Modify `Policy` class to accept multiple targets and deprecate previous single target methods
  - [x] Create `PropertyRegenerationPolicy` class for storing the policy in Java representation
  - [x] Create `PropertyRegenerationPolicyManager` for loading JSON policies into the `PropertyRegenerationPolicy` class
- [x] Policies will be read in by `ConfigLoader` where they will be reconciled and validated
  - [x] At most one policy per property
- [x] Update documentation to include information on utilizing policies

-------------Stretch Goal--------------------------

- [ ] Create `PropertyRegenerationRule` abstract class and basic rule classes that implement it:
  - [ ] Class to set a max number of uses before a property is regenerated
  - [ ] Class to set a expiration date before a property is regenerated
- [ ] Each rule within a policy will be initialized for each target
  - [ ] Metadata produced by each rule will be saved within the Vault property custom metadata
  - [ ] Create API call for [adding custom metadata](https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2#create-update-metadata)
- [ ] Switch to updated fork of [Vault Java Driver](https://github.com/jopenlibs/vault-java-driver)
  - [ ] Open PR to merge metadata call into the fork

#### ConfigLoader:
Adding to `foundation/foundation-configuration-store/src/test/resources/specifications/load-configurations.feature`
```
  Scenario: The ConfigLoader loads base and environment policies successfully

  Scenario: The ConfigLoader creates rule metadata for each rule in each policy

  Scenario: The policy is misformatted

  Scenario: The policies are not distinguishable

  Scenario: A policy has a undefined target property

  Scenario: A property has multiple policies

  Scenario: A policy has a rule implementation class could not be found
```
#### RuleMetadataDAO:
Creating new feature file at `extensions/extensions-configuration-store-vault/src/test/resources/specifications/`
```
  Scenario: The ConfigLoader can write rule metadata to the Vault server
    Given the policies are loaded, reconciled, and validated
    When rule metadata is created for each rule
    Then the rule metadata is saved to the Vault property custom metadata
```
## Test Strategy/Script
* **OTS only**
  * Build the following modules in aiSSEMBLE:
  ```
  mvn clean install -pl :aissemble-foundation-core-python,:foundation-core-java -Dmaven.build.cache.skipCache=true
  mvn clean install -pl :extensions-configuration-store-vault,:foundation-configuration-store -Dmaven.build.cache.skipCache=true -Pintegration-test
  ```

* Create a downstream project from the archetype using the following command:
```
mvn archetype:generate -DarchetypeGroupId=com.boozallen.aissemble -DarchetypeArtifactId=foundation-archetype -DarchetypeVersion=1.8.0-SNAPSHOT -DgroupId=com.example -DartifactId=example -DprojectGitUrl=url -DprojectName=example && cd example
```

* Add the attached [PysparkPipeline.json](https://github.com/user-attachments/files/15838614/PysparkPipeline.json) and [SparkPipeline.json](https://github.com/user-attachments/files/15838615/SparkPipeline.json) to `example-pipeline-models/src/main/resources/pipelines/`

* Run the following until all the manual actions are complete
```
mvn clean install
```

* Unzip the attached [example-configuration-store-docker.zip](https://github.com/user-attachments/files/15840520/example-configuration-store-docker.zip) and add it to the `example-docker/` folder

* Add the module to the `example-docker/pom.xml` file:
```
<module>example-configuration-store-docker</module> 
```

* Unzip the attached [config-store-service.zip](https://github.com/user-attachments/files/15840500/config-store-service.zip) and add it to the `example-deploy/src/main/resources/apps/` folder
* Add the following to your Tiltfile:
```
yaml = helm(
   'example-deploy/src/main/resources/apps/config-store-service',
   name='config-store-service',
   values=['example-deploy/src/main/resources/apps/config-store-service/values.yaml',
       'example-deploy/src/main/resources/apps/config-store-service/values-dev.yaml']
)
k8s_yaml(yaml)

# configuration-store
docker_build(
    ref='boozallen/example-configuration-store-docker',
    context='example-docker/example-configuration-store-docker',
    build_args=build_args,
    dockerfile='example-docker/example-configuration-store-docker/src/main/resources/docker/Dockerfile'
)
```

* Create the `policy-configuration.properties` file at `example-docker/example-spark-worker-docker/src/main/resources/krausening/base/` and add the following:
```
policies-location=/tmp/policies
```

* Create the the directory `example-docker/example-spark-worker-docker/src/main/resources/policies` and add the attached [policy-1.json](https://github.com/user-attachments/files/15840116/policy-1.json) and [policy-2.json](https://github.com/user-attachments/files/15840120/policy-2.json) to it

* Add the following to line 33 of `example-docker/example-spark-worker-docker/src/main/resources/docker/Dockerfile`:
```
COPY --chown=spark ./src/main/resources/policies/. /tmp/policies/
```

* **OTS only**
  * Copy the `aissemble_foundation_core_python-1.8.0.dev0-py3-none-any.whl` from the aiSSEMBLE repo to `example-docker/example-spark-worker-docker/target/dockerbuild/wheels/`

* Update the `execute_step_impl(self)` method in `example-pipelines/pyspark-pipeline/src/pyspark_pipeline/step/example_sync_step.py` to the following:
```
    def execute_step_impl(self) -> None:
        """
        This method performs the business logic of this step.
        """
        policyManager = DefaultPolicyManager.getInstance()

        ExampleSyncStep.logger.info(str(policyManager.getPolicy("policy-2").target))
        ExampleSyncStep.logger.info(str(policyManager.getPolicy("policy-2").targets))
        ExampleSyncStep.logger.info(str(policyManager.getPolicy("policy-1").targets))
``` 

* Add the following import to the top of the file:
```
from policy_manager import DefaultPolicyManager
```

* Update the `executeStepImpl() ` method in `example-pipelines/spark-pipeline/src/main/java/com/example/ExampleSyncStep.java` to the following:
```
    /**
     * {@inheritDoc}
     */
    @Override
    protected void executeStepImpl() {
        // TODO: Add your business logic here for this step!
        DefaultPolicyManager policyManager = DefaultPolicyManager.getInstance();

        logger.info(policyManager.getPolicy("policy-2").getTarget().toString());
        logger.info(policyManager.getPolicy("policy-2").getTargets().toString());
        logger.info(policyManager.getPolicy("policy-1").getTargets().toString());
    }
```

* Add the following import to the top of the file:
```
import com.boozallen.aissemble.core.policy.configuration.policymanager.DefaultPolicyManager;
```

* Run the following once: `mvn clean install -Dmaven.build.cache.skipCache=true`
* `tilt up vault`
* Once vault is ready, scroll up to find the root key, unseal keys
* In the `example-docker/example-configuration-store-docker/src/main/resources/krausening/base/encrypt.properties` file, update the `secrets.root.key` with the `ROOT KEY` content and `secrets.unseal.keys` with the `UNSEAL KEYS` content (make sure to remove the [], ", and any spaces)
* `tilt down`
* `tilt up`
* Verify you see the following logs in `config-store-service`:
```
INFO  com.boozallen.aissemble.configuration.store.ConfigLoader.loadPolicies():129 - Loaded 3 base policies and 2 environment policies
INFO  com.boozallen.aissemble.configuration.store.ConfigLoader.loadPolicies():133 - Using 3 policies after reconciling environment against base
INFO  com.boozallen.aissemble.configuration.store.ConfigLoader.validatePolicies():198 - Policy validation complete
```

* Run the `spark-pipeline` resource and verify you see the following logs:
```
WARN PolicyInput: Detected use of deprecated Json Property 'target'. Existing values should be moved to the new Json Property 'targets'.
WARN DefaultPolicy: Detected use of deprecated method 'getTarget()'. Existing method calls should be moved to the new method 'getTargets()'.
INFO ExampleSyncStep: com.boozallen.aissemble.core.policy.configuration.policy.Target@7af4286
INFO ExampleSyncStep: [com.boozallen.aissemble.core.policy.configuration.policy.Target@7af4286]
INFO ExampleSyncStep: [com.boozallen.aissemble.core.policy.configuration.policy.Target@41a6ab50, com.boozallen.aissemble.core.policy.configuration.policy.Target@6dc98c02]
```

* Run the `pyspark-pipeline` resource and verify you see the following logs:
```
WARNING PolicyInput: Detected use of deprecated Json Property 'target'. Existing values should be moved to the new Json Property 'targets'.
WARNING Policy: Detected use of deprecated attribute 'target'. Existing usage should be moved to the new attribute 'targets'.
INFO ExampleSyncStep: retrieve_url='/model-training-api/AWS_SECRET_ACCESS_KEY' type=None
INFO ExampleSyncStep: [Target(retrieve_url='/model-training-api/AWS_SECRET_ACCESS_KEY', type=None)]
INFO ExampleSyncStep: [Target(retrieve_url='/model-training-api/AWS_SECRET_ACCESS_KEY', type=None), Target(retrieve_url='/model-training-api/AWS_ACCESS_KEY_ID', type=None)]
```

* `tilt down`
## References/Additional Context
Add any other context, links, or screenshots about the feature request here.

**Comment by carter-cundiff** [comment]
DOD with @nartieri @csun-cpointe

**Comment by carter-cundiff** [comment]
OTS @peter-mcclonski

**Comment by ewilkins-csi** [comment]
Final Test completed ✅ 
Config Service logs:
![Screenshot 2024-06-18 at 11 11 24 AM](https://github.com/boozallen/aissemble/assets/80470879/30413bb1-5141-4548-95b2-58a68de24ecc)
Spark Pipeline:
![Screenshot 2024-06-18 at 11 28 11 AM](https://github.com/boozallen/aissemble/assets/80470879/eeb5b47b-b6f8-4327-94c4-cae76a4a49ea)
PySpark Pipeline:
![Screenshot 2024-06-18 at 11 30 19 AM](https://github.com/boozallen/aissemble/assets/80470879/2b4ae138-35dd-4227-abad-8c05a3daa916)
